packages/@react-aria/select/src/useSelect.ts

```diff
@@ -89,31 +89,45 @@ export function useSelect<T>(props: AriaSelectOptions<T>, state: SelectState<T>,
     ref
   );
 
+  let requestRef = useRef<number>();
+
   let onKeyDown = (e: KeyboardEvent) => {
+    let {target} = e;
+    let key: Key;
+
     switch (e.key) {
       case 'ArrowLeft': {
         // prevent scrolling containers
         e.preventDefault();
 
-        let key = state.selectedKey != null ? delegate.getKeyAbove(state.selectedKey) : delegate.getFirstKey();
-        if (key) {
-          state.setSelectedKey(key);
-        }
+        key = state.selectedKey != null ? delegate.getKeyAbove(state.selectedKey) : delegate.getFirstKey();
         break;
       }
       case 'ArrowRight': {
         // prevent scrolling containers
         e.preventDefault();
 
-        let key = state.selectedKey != null ? delegate.getKeyBelow(state.selectedKey) : delegate.getFirstKey();
-        if (key) {
-          state.setSelectedKey(key);
-        }
+        key = state.selectedKey != null ? delegate.getKeyBelow(state.selectedKey) : delegate.getFirstKey();
         break;
       }
     }
+
+    if (key) {
+      // Wait until the next render to change the selection, so that any other key handlers are done running.
+      requestRef.current = requestAnimationFrame(() => {
+        // If focus is no longer on the event target, don't change the selection.
+        if (target !== getOwnerDocument(target as Element).activeElement) {
+          return;
+        }
+        state.setSelectedKey(key);
+      });
```

Comment:
was this RAF the only way to get this to work? Did you try having table capture/handle all keyboard events (at least arrow key navigation)?

Reply:
TLDR; I tried a bunch of things to try to correct the order of operations so that a keyboard event on the table wouldn't propagate to the picker in the cell, but nothing worked.

`useGridCell` handles keyboard events using `onKeyDownCapture` and calls `stopPropagation`: 
https://github.com/adobe/react-spectrum/blob/fd3f09c798da23bf2148ba11f65e8856f8af9de6/packages/%40react-aria/grid/src/useGridCell.ts#L107C1-L245C6

`useSelectableCollection` handles the event in the bubble phase and does not `stopPropagation`: https://github.com/adobe/react-spectrum/blob/fd3f09c798da23bf2148ba11f65e8856f8af9de6/packages/%40react-aria/selection/src/useSelectableCollection.ts#L123-L291

`useSelect` seems to move events returned from the `useTypeSelect` from the capture phase to the bubble phase [here](https://github.com/adobe/react-spectrum/pull/5878/files/5ea877f566df0d8fbd9543963778c4c21bfa0f10#diff-7e599b9da08764665cb17b548a888c3caaa9a9fb472d331de37619fec7f3318eL133-L134): 
  ```ts
  typeSelectProps.onKeyDown = typeSelectProps.onKeyDownCapture;
  delete typeSelectProps.onKeyDownCapture;
  ```

Reply:
urgh, yup this is bringing back memories of edit mode pain points as well with all the differences in when components each handle the keydown handlers. I'll try to think on this some more, rather not have the RAF if at all possible, maybe we should pick up edit mode again...